packages/repository/src/repositories/legacy-juggler-bridge.ts

```diff
@@ -528,6 +534,50 @@ export class DefaultCrudRepository<
     return includeRelatedModels<T, Relations>(this, entities, include, options);
   }
 
+  /** Converts an entity object to a JSON object to check if it contains navigational property.
+   * Throws an error if `entity` contains navigational property.
+   * Also removed id property  for replaceById operation (mongo use case).
+   *
+   * @param entity
+   * @param options when attributw replacement is set to true, delete the id property to operate
+   * replaceById method.
+   */
+  protected ensurePersistable<R extends T>(
+    entity: R | DataObject<R>,
+    options: {replacement?: boolean; idProperty?: string} = {},
+  ): legacy.ModelData<legacy.PersistedModel> {
+    // FIXME(bajtos) Ideally, we should call toJSON() to convert R to data object
+    // Unfortunately that breaks replaceById for MongoDB connector, where we
+    // would call replaceId with id *argument* set to ObjectID value but
+    // id *property* set to string value.
+    /*
+    const data: AnyObject =
+      typeof entity.toJSON === 'function' ? entity.toJSON() : {...entity};
+    */
+
+    // proposal solution from agnes: delete the id property in the target dtat
+    // when runs replaceById
+
+    const data: AnyObject = new this.entityClass(entity);
+
+    const def = this.entityClass.definition;
+    for (const r in def.relations) {
+      const relName = def.relations[r].name;
+      if (relName in data) {
+        throw new Error(
+          `Navigational properties are not allowed in model data (model "${this.entityClass.modelName}" property "${relName}")`,
+        );
+      }
+    }
+    if (options.replacement === true) {
```

Comment:
Hmm, why do we need to deal with the `idProperty` in `ensurePersistable`? I don't think for replacement people can include the id from input, we should prevent the id field by excluding it from the request body schema defined in controller. 

See [the type used for request body](https://github.com/strongloop/loopback-next/blob/8a7f78dccdb384d19f16d382aaf0ac74be874e84/examples/todo/src/controllers/todo.controller.ts#L102), I think the solution should be changing it to sth like
```ts
@requestBody({
      content: {
        'application/json': {
          schema: getModelSchemaRef(Todo, {title: 'NewTodo', exclude: ['id']}),
        },
      },
    })
    todo: Omit<Todo, 'id'>,
```

Reply:
I just tested out the code with @bajtos 's fix of `ObjectId`: https://github.com/strongloop/loopback-connector-mongodb/pull/553. I think I can remove the `options.replacement` part cause it's fixed 👏 

And yes I agree that `id` should be excluded from `update/replace` methods at **both** controller and repository level. Currently `update/replacement` methods take `id` property and ignore it in the operation. The behavior is **correct** but I think it should throw errors when `id` is included. 

So I am thinking, like Miroslav suggested above, the new function can perform validations such as checking nav properties, and checking if id is included for `update/replace`.

Reply:
> I just tested out the code with @bajtos 's fix of ObjectId: strongloop/loopback-connector-mongodb#553. I think I can remove the options.replacement part cause it's fixed 👏

Awesome!

> And yes I agree that id should be excluded from update/replace methods at both controller and repository level. Currently, update/replacement methods take id property and ignore it in the operation. The result is correct but I think it should throw errors when id is included.

I believe it was an intentional design to require `id` property to be present in the payload of `replaceById` request. This operation does exactly what is says - replaces the entire model instance in the database with the supplied data. This is important when free-form properties come to play, because it's not possible to remove a property via `patch` calls. The call `patchById(id, {someProp: undefined})` is a no-op, because `undefined` cannot be represented in JSON.

I feel it's also better UX when the `id` is allowed in the model data sent to both `patch` and `replace` methods. Clients often store the model instance in a variable, then make some change to some properties (e.g. via UI) and then want to persist that changes. If the PK property is forbidden in the payload, then the clients must do extra work to remove it.